napari/_vispy/vendored/volume.py

```diff
@@ -379,11 +388,69 @@
     iproj.z /= iproj.w;
     gl_FragDepth = (iproj.z+1.0)/2.0;
     */
-}}
+}
 
 
 """  # noqa
 
+RAYCASTING_SETUP_VOLUME = """
+    // Compute the distance to the front surface or near clipping plane
+    float distance = dot(nearpos-v_position, view_ray);
+    distance = max(distance, min((-0.5 - v_position.x) / view_ray.x,
+                            (u_shape.x - 0.5 - v_position.x) / view_ray.x));
+    distance = max(distance, min((-0.5 - v_position.y) / view_ray.y,
+                            (u_shape.y - 0.5 - v_position.y) / view_ray.y));
+    distance = max(distance, min((-0.5 - v_position.z) / view_ray.z,
+                            (u_shape.z - 0.5 - v_position.z) / view_ray.z));
+
+    // Now we have the starting position on the front surface
+    vec3 front = v_position + view_ray * distance;
+
+    // Decide how many steps to take
+    int nsteps = int(-distance / u_relative_step_size + 0.5);
+    float f_nsteps = float(nsteps);
+    if( nsteps < 1 )
+        discard;
+
+    // Get starting location and step vector in texture coordinates
+    vec3 step = ((v_position - front) / u_shape) / f_nsteps;
+    vec3 start_loc = front / u_shape;
+"""
+
+RAYCASTING_SETUP_PLANE = """
+    // find intersection of view ray with plane in data coordinates
+    vec3 intersection = intersectLinePlane(v_position.xyz, view_ray, 
+                                           u_plane_position, u_plane_normal);
+    // and texture coordinates
+    vec3 intersection_tex = intersection / u_shape;
+
+    // discard if intersection not in texture
+    if (intersection_tex.x > 1.0 )
+        discard;
+    if (intersection_tex.y > 1.0 )
+        discard;
+    if (intersection_tex.z > 1.0 )
+        discard;
+    if (intersection_tex.x < 0.0 )
+        discard;
+    if (intersection_tex.y < 0.0 )
+        discard;
+    if (intersection_tex.z < 0.0 )
+        discard;
```

Comment:
If you wanted to reduce the branching here, I think you could do something like:

```
float out_of_bounds = 0;

out_of_bounds += float(intersection_tex.x > 1);

// do the rest of the comparisons

if (out_of_bounds > 0){
    discard;
}
```

Reply:
super nice! thanks